File: GLLara/GLLItem+MeshExport.m
//
//  GLLItem+MeshExport.m
//  GLLara
//
//  Created by Torsten Kammer on 25.02.13.
//  Copyright (c) 2013 Torsten Kammer. All rights reserved.
//

#import "GLLItem+MeshExport.h"

#import "GLLItemBone.h"
#import "GLLItemMesh+MeshExport.h"
#import "TROutDataStream.h"

#import "GLLara-Swift.h"

@implementation GLLItem (MeshExport)

- (NSData *)writeBinaryError:(NSError *__autoreleasing*)error;
{
    TROutDataStream *stream = [[TROutDataStream alloc] init];
    
    [stream appendUint32:(uint32_t) self.bones.count];
    for (GLLItemBone *bone in self.bones)
        [stream appendData:[bone.bone writeBinary]];
    
    NSArray *toExport = [self.meshes.array filteredArrayUsingPredicate:[NSPredicate predicateWithFormat:@"shouldExport == YES"]];
    [stream appendUint32:(uint32_t) toExport.count];
    for (GLLItemMesh *mesh in toExport)
    {
        NSData *meshData = [mesh writeBinaryError:error];
        if (!meshData) return nil;
        [stream appendData:meshData];
    }
    
    return stream.data;
}
- (NSString *)writeASCIIError:(NSError *__autoreleasing*)error;
{
    NSMutableString *string = [NSMutableString string];
    
    [string appendFormat:@"%lu\n", self.bones.count];
    for (GLLItemBone *bone in self.bones)
        [string appendFormat:@"%@\n", [bone.bone writeASCII]];
    
    NSArray *toExport = [self.meshes.array filteredArrayUsingPredicate:[NSPredicate predicateWithFormat:@"shouldExport == YES"]];
    [string appendFormat:@"%lu\n", toExport.count];
    for (GLLItemMesh *mesh in toExport)
    {
        NSString *meshString = [mesh writeASCIIError:error];
        if (!meshString) return nil;
        [string appendFormat:@"%@\n", meshString];
    }
    
    return string;
}

@end

File: GLLara/GLLItem+MeshExport.swift
//
//  GLLItem+MeshExport.swift
//  GLLara
//
//  Created by Torsten Kammer on 25.02.13.
//  Copyright (c) 2013 Torsten Kammer. All rights reserved.
//

import Foundation

extension GLLItem {
    
    @objc func writeBinary() throws -> Data {
        let stream = TROutDataStream()
        
        stream.appendUint32(UInt32(bones.count))
        for bone in bones {
            stream.appendData((bone as! GLLItemBone).bone.writeBinary())
        }
        
        var count: UInt32 = 0
        for mesh in meshes {
            if (mesh as! GLLItemMesh).shouldExport {
                count += 1
            }
        }
        stream.appendUint32(count)
        for mesh in meshes {
            let itemMesh = mesh as! GLLItemMesh
            if itemMesh.shouldExport {
                try stream.appendData(itemMesh.writeBinary())
            }
        }
        
        return stream.data()!;
    }
    
    @objc func writeASCII() throws -> String {
        var string = ""
        
        string += "\(bones.count)\n"
        for bone in bones {
            string += (bone as! GLLItemBone).bone.writeASCII()
            string += "\n"
        }
        
        var count = 0
        for mesh in meshes {
            if (mesh as! GLLItemMesh).shouldExport {
                count += 1
            }
        }
        string += "\(count)\n"
        for mesh in meshes {
            let itemMesh = mesh as! GLLItemMesh
            if itemMesh.shouldExport {
                string += try itemMesh.writeASCII()
                string += "\n"
            }
        }
        
        return string;
    }
    
}
